File: hanghae99/java/SmallestInfiniteSet.java
package hanghae99.java;

import java.util.PriorityQueue;

public class SmallestInfiniteSet {
    PriorityQueue<Integer> pq;
    int i = 1;

    public SmallestInfiniteSet() {
        pq = new PriorityQueue<>();
    }

    public int popSmallest() {
        if (pq.isEmpty()) {
            return i++;
        } else {
            return pq.poll();
        }
    }

    public void addBack(int num) {
        if (num < i && !pq.contains(num)) {
            pq.add(num);
        }
    }
}

File: hanghae99/kotlin/SmallestInfiniteSet.kt
package hanghae99.kotlin

import java.util.*

class SmallestInfiniteSet {
	var pq: PriorityQueue<Int>? = null
	var i: Int = 1

	fun SmallestInfiniteSet() {
		pq = PriorityQueue()
	}

	fun popSmallest(): Int {
		return if (pq != null && pq!!.isEmpty()) {
			i++
		} else {
			pq!!.poll()
		}
	}

	fun addBack(num: Int) {
		if (num < i && !pq!!.contains(num)) {
			pq!!.add(num)
		}
	}

}